### PR TITLE
Import test: improve docs, add more tests

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Follow-up to #72497. Improves documentation of import sanity test, 

Also hopefully improves the situation in #73596, CC @rhysmeister.

##### ISSUE TYPE
- Docs Pull Request
- Test Pull Request

##### COMPONENT NAME
ansible-test import sanity test

### CODE REVIEW

**Comment by felixfontein** [comment]
ready_for_review

**Review by evgeni — APPROVED** [review]

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst [hunk 1]
```diff
@@ -1,5 +1,39 @@
...
+All Python imports in the directories specified that are not from a limited set of possibilities listed below
+must be imported in a try/except ImportError block as follows:
+
+.. code-block:: python
+
+    # Instead of 'import another_library', do:
+
+    try:
+        import another_library
+        HAS_ANOTHER_LIBRARY = True
```
Comment (webknjaz) [inline docs/docsite/rst/dev_guide/testing/sanity/import.rst]: For readability, it's probably better to have this in the `else:` block.

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst [hunk 2]
```diff
@@ -1,5 +1,39 @@
...
+    # Later in module code:
+
+        if not HAS_ANOTHER_LIBRARY:
+            # Needs: from ansible.module_utils.basic import missing_required_lib
+            module.fail_json(msg=missing_required_lib('another_library'))
+
+    # Later in plugin code:
+
+        if not HAS_ANOTHER_LIBRARY:
+            raise AnsibleError('another_library must be installed to use this plugin')
```
Comment (webknjaz) [inline docs/docsite/rst/dev_guide/testing/sanity/import.rst]: I want to add that if the code targets a new enough Python 3 version, the example could preserve more metadata about the exception causes:
```python
    try:
        import another_library
    except ImportError as imp_exc:
        another_library_import_exc = imp_exc
    else:
        another_library_import_exc = None


    # Later in module code:

        if another_library_import_exc is not None:
            # Needs: from ansible.module_utils.basic import missing_required_lib
            module.fail_json(msg=missing_required_lib('another_library'))

    # Later in plugin code:

        if another_library_import_exc is not None:
            raise AnsibleError('another_library must be installed to use this plugin') from another_library_import_exc
```

For older Pythons the example could use `six.raise_from()` shim.

File: test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+author:
+  - Ansible Core Team
+'''
+
+EXAMPLES = '''
+- debug:
+    msg: "{{ lookup('ns.col.bad') }}"
+'''
+
+RETURN = ''''''
```
Comment (webknjaz) [inline test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py]: How about 
```suggestion
RETURN = ''
```

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst (hunk 1 above)
Comment (felixfontein) [reply docs/docsite/rst/dev_guide/testing/sanity/import.rst]: Fixed in 4f9e1277be264add28921b348cdaf820b81a2a9f.

File: test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py (hunk 3 above)
Comment (felixfontein) [reply test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py]: I copied that from existing tests in the same collection. I've changed it to `''' # '''` as recommended here: https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block (8d8ecba47cfda2b94a958ef584d426a45e327db1).

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst (hunk 2 above)
Comment (felixfontein) [reply docs/docsite/rst/dev_guide/testing/sanity/import.rst]: I split up the example into modules vs. plugins, and used `traceback` to return the import traceback in `module.fail_json()`, and `six.raise_from()` for plugins (e74bec06995093c876c6fd81bc5615951641cba8).

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst [hunk 4]
```diff
@@ -1,5 +1,69 @@
...
+
+       try:
+           import another_library
+       except ImportError as imp_exc:
+           ANOTHER_LIBRARY_IMPORT_ERROR = imp_exc
+       else:
+           ANOTHER_LIBRARY_IMPORT_ERROR = None
+
+
+       # Later in plugin code:
```
Comment (Andersson007) [inline docs/docsite/rst/dev_guide/testing/sanity/import.rst]:
```suggestion
       # Later in plugin code:
```
let's specify a suitable place maybe, e.g. in `__init__`?:)

**Comment by Andersson007** [comment]
and thank you for the clarification! it helps much

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst (hunk 4 above)
Comment (felixfontein) [reply docs/docsite/rst/dev_guide/testing/sanity/import.rst]: @Andersson007 how about:
```suggestion
       # Later in plugin code, for example in __init__ of the plugin:
```

**Comment by felixfontein** [comment]
CC @mattclay @acozine @lmalivert @samccann

**Review by mattclay — APPROVED** [review]

**Review by Andersson007 — APPROVED** [review]

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst [hunk 5]
```diff
@@ -1,5 +1,69 @@
 import
 ======
 
-All Python imports in ``lib/ansible/modules/`` and ``lib/ansible/module_utils/`` which are not from the Python standard library
-must be imported in a try/except ImportError block.
+All Python imports in the directories specified that are not from a limited set of possibilities listed below
+must be imported in a try/except ImportError block as follows:
```
Comment (acozine) [inline docs/docsite/rst/dev_guide/testing/sanity/import.rst]: Without the specific directories, I find this a little confusing. It's also passive voice. What about this?
```suggestion
Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. Import all other Python libraries in a try/except ImportError block as follows:
```
Adding a short sentence about why this is useful/important/required would help make the rule more memorable.

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst [hunk 6]
```diff
@@ -1,5 +1,69 @@
...
+       # Later in plugin code, for example in __init__ of the plugin:
+
+       if ANOTHER_LIBRARY_IMPORT_ERROR:
+           raise_from(
+               AnsibleError('another_library must be installed to use this plugin'),
+               ANOTHER_LIBRARY_IMPORT_ERROR)
+           # If you target only newer Python 3 versions, you can also use the
+           # 'raise ... from ...' syntax.
+
+The following shows which unchecked imports are allowed from which directories:
```
Comment (acozine) [inline docs/docsite/rst/dev_guide/testing/sanity/import.rst]:
```suggestion
Ansible allows the following unchecked imports from these specific directories:
```

File: test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+name: bad
```
Comment (acozine) [inline test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py:7]: I love this! Bad lookup! Bad!

**Review by acozine — COMMENTED** [review]
Added a couple of suggestions for clarity.

File: test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py (hunk 7 above)
Comment (felixfontein) [reply test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py:7]: I copied that name from an existing module in that test suite: https://github.com/ansible/ansible/blob/916d12eb13fba6b2ea1870d133599cfdc1b1cb98/test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/modules/bad.py

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst (hunk 5 above)
Comment (felixfontein) [reply docs/docsite/rst/dev_guide/testing/sanity/import.rst]: @acozine I added an explanation as well: 11d2f5f2a5ecbd6e3f117c7b7580597bc3751123

File: docs/docsite/rst/dev_guide/testing/sanity/import.rst [hunk 8]
```diff
@@ -1,5 +1,68 @@
 import
 ======
 
-All Python imports in ``lib/ansible/modules/`` and ``lib/ansible/module_utils/`` which are not from the Python standard library
-must be imported in a try/except ImportError block.
+Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. This allows Ansible to give better error messages to the user, and is a requirement for other sanity tests (such as ``validate-modules``). Import all other Python libraries in a try/except ImportError block as follows:
```
Comment (acozine) [inline docs/docsite/rst/dev_guide/testing/sanity/import.rst]:
```suggestion
Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. Import all other Python libraries in a try/except ImportError block to support sanity tests such as ``validate-modules`` and to allow Ansible to give better error messages to the user. To import a library in a try/except ImportError block:
```

**Review by acozine — APPROVED** [review]
Thanks for adding the reason for using try/except blocks. I'd move things around a bit just so the comparison of unchecked imports and "all other Python libraries" is easier to follow.

**Review by mattclay — APPROVED** [review]

**Comment by felixfontein** [comment]
@evgeni @webknjaz @mattclay @Andersson007 @acozine thanks for all your reviews, suggestions, and for merging this!

**Pull request merged**